***Marcin Brzózka  
nr albumu 405499  
grupa 2***

***Połączenie z bazą danych przy pomocy sqlAlchemy***

In [3]:
from sqlalchemy import create_engine
import pandas as pd

# db_string = "postgres://postgres:postgres@localhost:5432/test"
db_string = "postgresql://wbauer_adb:adb2020@pgsql-196447.vipserv.org:5432/wbauer_adb"

db = create_engine(db_string)
connection_sqlalchemy = db.connect()
connection = db.connect()

Zapytanie do bazy i wypisanie wyników

In [4]:
# result_set = db.execute("SELECT * FROM city")  
# for r in result_set:  
#     print(r)

Połączenie z bazą danych przy pomocy psycopg2 i pandas

In [5]:
# import psycopg2 as pg
# import pandas as pd

# connection = pg.connect(host='pgsql-196447.vipserv.org', port=5432, dbname='wbauer_adb', user='wbauer_adb', password='adb2020')

Zapytanie do bazy i wypisanie wyników

In [6]:
# df = pd.read_sql('select * from city',con=connection)
# df

**1. Ile kategorii filmów mamy w wypożyczalni?**

In [7]:
df = pd.read_sql_query('SELECT count(*) FROM category',con=connection)
df

count
0     16

Jest 16 kategorii filmowych.

**2. Wyświetl listę kategorii w kolejności alfabetycznej.**

In [8]:
df = pd.read_sql_query('SELECT name FROM category ORDER BY category ASC',con=connection)
df

name
0        Action
1     Animation
2      Children
3      Classics
4        Comedy
5   Documentary
6         Drama
7        Family
8       Foreign
9         Games
10       Horror
11        Music
12          New
13       Sci-Fi
14       Sports
15       Travel

**3. Znajdź najstarszy i najmłodszy film do wypożyczenia.**

In [9]:
df1 = pd.read_sql_query('SELECT title, release_year FROM film ORDER BY release_year ASC LIMIT 1',con=connection)
df2 = pd.read_sql_query('SELECT title, release_year FROM film ORDER BY release_year DESC LIMIT 1',con=connection)
print(df1)
print("\n\n")
print(df2)

             title  release_year
0  Chamber Italian          2006



             title  release_year
0  Chamber Italian          2006


Wszystkie filmy są z tego samego roku więc dlatego najmłodszy i najstarszy film wyszedł taki sam.

**4. Ile wypożyczeń odbyło się między 2005-07-01 a 2005-08-01?**

In [12]:
df = pd.read_sql_query("SELECT count(rental_date) FROM rental WHERE rental_date BETWEEN '2005-07-01 00:00:00'::timestamp AND '2005-08-01 23:59:59'::timestamp",con=connection)
df

count
0   7380

Wypożyczono 7380 filmów (mprzy uwzględnieniu wypożyczeń w pierwszy i ostatni dzień)

**5. Ile wypożyczeń odbyło się między 2010-01-01 a 2011-02-01?**

In [9]:
df = pd.read_sql_query("SELECT count(rental_date) FROM rental WHERE rental_date BETWEEN '2010-01-01 00:00:00'::timestamp AND '2011-02-01 23:59:59'::timestamp",con=connection)
df

count
0      0

Nie wypożyczono filmów w tym okresie lub brak danych.

**6. Znajdź największą płatność wypożyczenia.**

In [10]:
df = pd.read_sql_query("SELECT * FROM payment ORDER BY amount DESC LIMIT 1",con=connection)
df

payment_id  customer_id  staff_id  rental_id  amount  \
0       20403          362         1      14759   11.99   

                payment_date  
0 2007-03-21 21:57:24.996577

**7. Znajdź wszystkich klientów z Polski, Nigerii lub Bangladeszu.**

In [13]:
df = pd.read_sql_query("""
    SELECT customer.*, country FROM country 
    INNER JOIN city ON city.country_id = country.country_id
    INNER JOIN address ON address.city_id = city.city_id
    INNER JOIN customer ON customer.address_id = address.address_id
    WHERE country IN ('Poland', 'Nigeria', 'Bangladesh') ORDER BY customer_id ASC
    """,con=connection)
df

customer_id  store_id first_name  last_name  \
0            18         2      Carol     Garcia   
1            21         1   Michelle      Clark   
2            80         1    Marilyn       Ross   
3           103         1     Gladys   Hamilton   
4           128         1   Marjorie     Tucker   
5           156         1     Bertha   Ferguson   
6           198         2      Elsie     Kelley   
7           232         2  Constance       Reid   
8           250         2         Jo     Fowler   
9           270         1       Leah     Curtis   
10          277         2       Olga    Jimenez   
11          280         2     Tracey    Barrett   
12          284         1      Sonia    Gregory   
13          286         1      Velma      Lucas   
14          318         1      Brian      Wyman   
15          329         2      Frank   Waggoner   
16          332         1    Stephen     Qualls   
17          380         1    Russell    Brinson   
18          409         2     Rodney    Moeller   
19          501         1      Ruben      Geary   
20          558         1     Jimmie  Eggleston   
21          562         1    Wallace      Slone   
22          571         2    Johnnie   Chisholm   
23          572         1     Sidney   Burleson   

                                  email  address_id  activebool create_date  \
0       carol.garcia@sakilacustomer.org          22        True  2006-02-14   
1     michelle.clark@sakilacustomer.org          25        True  2006-02-14   
2       marilyn.ross@sakilacustomer.org          84        True  2006-02-14   
3    gladys.hamilton@sakilacustomer.org         107        True  2006-02-14   
4    marjorie.tucker@sakilacustomer.org         132        True  2006-02-14   
5    bertha.ferguson@sakilacustomer.org         160        True  2006-02-14   
6       elsie.kelley@sakilacustomer.org         202        True  2006-02-14   
7     constance.reid@sakilacustomer.org         236        True  2006-02-14   
8          jo.fowler@sakilacustomer.org         254        True  2006-02-14   
9        leah.curtis@sakilacustomer.org         275        True  2006-02-14   
10      olga.jimenez@sakilacustomer.org         282        True  2006-02-14   
11    tracey.barrett@sakilacustomer.org         285        True  2006-02-14   
12     sonia.gregory@sakilacustomer.org         289        True  2006-02-14   
13       velma.lucas@sakilacustomer.org         291        True  2006-02-14   
14       brian.wyman@sakilacustomer.org         323        True  2006-02-14   
15    frank.waggoner@sakilacustomer.org         334        True  2006-02-14   
16    stephen.qualls@sakilacustomer.org         337        True  2006-02-14   
17   russell.brinson@sakilacustomer.org         385        True  2006-02-14   
18    rodney.moeller@sakilacustomer.org         414        True  2006-02-14   
19       ruben.geary@sakilacustomer.org         506        True  2006-02-14   
20  jimmie.eggleston@sakilacustomer.org         564        True  2006-02-14   
21     wallace.slone@sakilacustomer.org         568        True  2006-02-14   
22  johnnie.chisholm@sakilacustomer.org         577        True  2006-02-14   
23   sidney.burleson@sakilacustomer.org         578        True  2006-02-14   

               last_update  active     country  
0  2013-05-26 14:49:45.738       1     Nigeria  
1  2013-05-26 14:49:45.738       1  Bangladesh  
2  2013-05-26 14:49:45.738       1     Nigeria  
3  2013-05-26 14:49:45.738       1     Nigeria  
4  2013-05-26 14:49:45.738       1      Poland  
5  2013-05-26 14:49:45.738       1     Nigeria  
6  2013-05-26 14:49:45.738       1     Nigeria  
7  2013-05-26 14:49:45.738       1     Nigeria  
8  2013-05-26 14:49:45.738       1     Nigeria  
9  2013-05-26 14:49:45.738       1      Poland  
10 2013-05-26 14:49:45.738       1     Nigeria  
11 2013-05-26 14:49:45.738       1     Nigeria  
12 2013-05-26 14:49:45.738       1     Nigeria  
13 2013-05-26 14:49:45.738       1     Nigeria  
14 2013-05-26 14:

**8. Gdzie mieszkają członkowie personelu?**

Tylko imię, nazwisko i adres:

In [13]:
df = pd.read_sql_query("""
    SELECT first_name, last_name, address FROM staff
    LEFT JOIN address ON staff.address_id = address.address_id
    """, con=connection)

df

first_name last_name               address
0       Mike   Hillyer     23 Workhaven Lane
1        Jon  Stephens  1411 Lillydale Drive

Imię, nazwisko, adres, miasto, kraj:

In [14]:
df = pd.read_sql("""SELECT first_name, last_name, address, city, country FROM country
                 INNER JOIN city ON country.country_id = city.country_id
                 INNER JOIN address ON city.city_id = address.city_id
                 INNER JOIN staff ON address.address_id = staff.address_id
                 ORDER BY staff_id ASC
                 """ ,con=connection)
df

first_name last_name               address        city    country
0       Mike   Hillyer     23 Workhaven Lane  Lethbridge     Canada
1        Jon  Stephens  1411 Lillydale Drive   Woodridge  Australia

**9. Ilu pracowników mieszka w Argentynie lub Hiszpanii?**

In [15]:
df = pd.read_sql("""SELECT count(*) FROM country
                 INNER JOIN city ON country.country_id = city.country_id
                 INNER JOIN address ON city.city_id = address.city_id
                 INNER JOIN staff ON address.address_id = staff.address_id
                 WHERE country IN ('Argentina', 'Spain')
                 """ ,con=connection)
df

count
0      0

**10. Jakie kategorie filmów zostały wypożyczone przez klientów?**

In [16]:
df = pd.read_sql("""SELECT DISTINCT name FROM category
                 INNER JOIN film_category ON film_category.category_id = category.category_id
                 INNER JOIN inventory ON inventory.film_id = film_category.film_id
                 INNER JOIN rental ON rental.inventory_id = inventory.inventory_id
                 ORDER BY name ASC
                 """ ,con=connection)
df

name
0        Action
1     Animation
2      Children
3      Classics
4        Comedy
5   Documentary
6         Drama
7        Family
8       Foreign
9         Games
10       Horror
11        Music
12          New
13       Sci-Fi
14       Sports
15       Travel

**11. Znajdź wszystkie kategorie filmów wypożyczonych w Ameryce.**

In [17]:
df = pd.read_sql("""SELECT DISTINCT name FROM category
                 INNER JOIN film_category ON film_category.category_id = category.category_id
                 INNER JOIN inventory ON film_category.film_id = inventory.film_id
                 INNER JOIN rental ON rental.inventory_id = inventory.inventory_id
                 INNER JOIN customer ON customer.customer_id = rental.customer_id
                 INNER JOIN address ON address.address_id = customer.address_id
                 INNER JOIN city ON city.city_id = address.city_id
                 INNER JOIN country ON country.country_id = city.country_id
                 WHERE country.country = 'United States'
                 ORDER BY name ASC
                 """ ,con=connection)
df

name
0        Action
1     Animation
2      Children
3      Classics
4        Comedy
5   Documentary
6         Drama
7        Family
8       Foreign
9         Games
10       Horror
11        Music
12          New
13       Sci-Fi
14       Sports
15       Travel

**12. Znajdź wszystkie tytuły filmów, w których grał: Olympia Pfeiffer lub Julia Zellweger lub Ellen Presley.**

In [18]:
df = pd.read_sql("""SELECT DISTINCT title FROM film
                 INNER JOIN film_actor ON film_actor.film_id = film.film_id
                 INNER JOIN actor ON actor.actor_id = film_actor.actor_id
                 WHERE (first_name = 'Olympia' AND last_name = 'Pfeiffer') 
                     OR (first_name = 'Julia' AND last_name = 'Zellweger') 
                     OR (first_name = 'Ellen' AND last_name = 'Presley')
                 ORDER BY title ASC
                 """ ,con=connection)
df

title
0            Badman Dawn
1        Bilko Anonymous
2   Breakfast Goldfinger
3      Caribbean Liberty
4       Casper Dragonfly
..                   ...
64             Turn Star
65            Wait Cider
66          Women Dorado
67             Won Dares
68         Wyoming Storm

[69 rows x 1 columns]